hydromt/data_adapter/dataset.py

```diff
@@ -0,0 +1,431 @@
+                mv = nodata.get(k, None)
+                if mv is not None and ds[k].attrs.get("_FillValue", None) is None:
+                    ds[k].attrs["_FillValue"] = mv
+        return ds
+
+    def _apply_unit_conversion(
+        self, ds: xr.Dataset, logger: logging.Logger = logger
+    ) -> xr.Dataset:
+        unit_names = list(self.unit_mult.keys()) + list(self.unit_add.keys())
+        unit_names = [k for k in unit_names if k in ds.data_vars]
+        if len(unit_names) > 0:
+            logger.debug(f"Convert units for {len(unit_names)} variables.")
+        for name in list(set(unit_names)):  # unique
+            m = self.unit_mult.get(name, 1)
+            a = self.unit_add.get(name, 0)
+            da = ds[name]
+            attrs = da.attrs.copy()
+            nodata_isnan = da.attrs.get("_FillValue", None) is None or np.isnan(
+                da.attrs.get("_FillValue", None)
+            )
+            # nodata value is explicitly set to NaN in case no nodata value is provided
+            nodata = np.nan if nodata_isnan else da.attrs["_FillValue"]
+            data_bool = ~np.isnan(da) if nodata_isnan else da != nodata
+            ds[name] = xr.where(data_bool, da * m + a, nodata)
+            ds[name].attrs.update(attrs)  # set original attributes
+        return ds
+
+    def _shift_time(
+        self, ds: xr.Dataset, logger: logging.Logger = logger
+    ) -> xr.Dataset:
+        dt = self.unit_add.get("time", 0)
+        return shift_dataset_time(dt=dt, ds=ds, logger=logger)
+
+    @staticmethod
+    def _slice_data(
+        ds: xr.Dataset,
+        variables: Optional[str | List[str]] = None,
+        time_tuple=Optional[Tuple[str] | Tuple[datetime]],
+        logger: logging.Logger = logger,
+    ) -> xr.Dataset:
+        """Slice the dataset in space and time.
+
+        Arguments
+        ---------
+        ds : xarray.Dataset or xarray.DataArray
+            The Dataset to slice.
+        variables : str or list of str, optional.
+            Names of variables to return.
+        time_tuple : tuple of str, datetime, optional
+            Start and end date of period of interest. By default the entire time period
+            of the dataset is returned.
+
+        Returns
+        -------
+        ds : xarray.Dataset
+            The sliced Dataset.
+        """
+        if isinstance(ds, xr.DataArray):
+            if ds.name is None:
+                ds.name = "data"
+            ds = ds.to_dataset()
+        elif variables is not None:
+            variables = np.atleast_1d(variables).tolist()
+            if len(variables) > 1 or len(ds.data_vars) > 1:
+                mvars = [var not in ds.data_vars for var in variables]
+                if any(mvars):
+                    raise ValueError(f"Dataset: variables not found {mvars}")
+                ds = ds[variables]
+        if time_tuple is not None:
+            ds = DatasetAdapter._slice_temporal_dimension(ds, time_tuple, logger=logger)
```

Comment:
This will fail I think if the slice would return an empty dataset because of the underlying code. This would be a good place to apply the new error pattern